src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -40,12 +40,48 @@ internal Task<ManagedIdentityResponse> SendTokenRequestForManagedIdentityAsync(A
         // This method tries to create managed identity source for different sources, if none is created then defaults to IMDS.
         private static AbstractManagedIdentity SelectManagedIdentitySource(RequestContext requestContext)
         {
-            return 
-                ServiceFabricManagedIdentitySource.TryCreate(requestContext) ??
-                AppServiceManagedIdentitySource.TryCreate(requestContext) ?? 
-                CloudShellManagedIdentitySource.TryCreate(requestContext) ??
-                AzureArcManagedIdentitySource.TryCreate(requestContext) ??
-                new ImdsManagedIdentitySource(requestContext);
+            return s_managedIdentitySourceDetected.Value switch
+            {
+                ManagedIdentitySource.ServiceFabric => ServiceFabricManagedIdentitySource.TryCreate(requestContext),
+                ManagedIdentitySource.AppService => AppServiceManagedIdentitySource.TryCreate(requestContext),
+                ManagedIdentitySource.CloudShell => CloudShellManagedIdentitySource.TryCreate(requestContext),
+                ManagedIdentitySource.AzureArc => AzureArcManagedIdentitySource.TryCreate(requestContext),
```

Comment:
Nit: I guess all those `TryCreate()` can be renamed to `Create()` now. Not a big deal, though, I'm approving.

Reply:
I think we can simply call the constructor now. Thanks for pointing this out.